application/src/main/java/org/opentripplanner/transit/service/DefaultTransitService.java

```diff
@@ -91,6 +96,45 @@ public DefaultTransitService(
     this.timetableSnapshot = timetableSnapshotBuffer;
   }
 
+  public Optional<List<TripTimeOnDate>> getScheduledTripTimes(Trip trip) {
+    TripPattern tripPattern = getPatternForTrip(trip);
+    if (tripPattern == null) {
+      return Optional.empty();
+    }
+    return Optional.ofNullable(
+      TripTimeOnDate.fromTripTimes(tripPattern.getScheduledTimetable(), trip)
+    );
+  }
+
+  public Optional<List<TripTimeOnDate>> getTripTimeOnDates(Trip trip, LocalDate serviceDate) {
+    TripPattern pattern = getPatternForTrip(trip, serviceDate);
+
+    Timetable timetable = getTimetableForTripPattern(pattern, serviceDate);
+
+    // If realtime moved pattern back to original trip, fetch it instead
+    if (timetable.getTripIndex(trip.getId()) == -1) {
+      LOG.warn(
+        "Trip {} not found in realtime pattern. This should not happen, and indicates a bug.",
+        trip
+      );
+      pattern = getPatternForTrip(trip);
+      timetable = getTimetableForTripPattern(pattern, serviceDate);
+    }
+
+    // This check is made here to avoid changing TripTimeOnDate.fromTripTimes
+    TripTimes times = timetable.getTripTimes(trip);
+    if (!this.getServiceCodesRunningForDate(serviceDate).contains(times.getServiceCode())) {
+      return Optional.empty();
+    } else {
+      Instant midnight = ServiceDateUtils
+        .asStartOfService(serviceDate, this.getTimeZone())
+        .toInstant();
+      return Optional.ofNullable(
+        TripTimeOnDate.fromTripTimes(timetable, trip, serviceDate, midnight)
+      );
+    }
+  }
+
```

Comment:
I wanted to move the implementation above into a "delegate", but I see now that that is a bit difficult without creating circular dependencies - so I will not request this for this PR. We should instead try to do this as part of the transit model refactoring. There should be a service responsible timetable data, but we do not have that - the logic is split into several places.